### PR TITLE
Add draft for dask-cudf

### PR DESCRIPTION
**Pull request by mrocklin**
This post talks about Dask and cuDF and early plans to make them work well
together.

### CODE REVIEW

File: _posts/2019-01-13-dask-cudf-first-steps.md [hunk 1]
```diff
@@ -0,0 +1,437 @@
...
+4.  We had to make Dask Dataframe a bit more flexible and assume less about its
+    constituent dataframes being exactly Pandas dataframes.  (see
+    [dask/dask #4359](https://github.com/dask/dask/pull/4359) and
+    [dask/dask #4375](https://github.com/dask/dask/pull/4375) for examples).
+    I suspect that there will by many more small changes like
+    these necessary in the future.
+
+These problems are representative of dozens more similar issues.  They are are
+all fixable and indeed, many are actively being fixed today by the [good folks
+working on RAPIDS](https://github.com/rapidsai/cudf/graphs/contributors).
```
Comment (kkraus14) [inline _posts/2019-01-13-dask-cudf-first-steps.md:277]: Note because of the branch-0.6 change this graph is empty.

File: _posts/2019-01-13-dask-cudf-first-steps.md [hunk 2]
```diff
@@ -0,0 +1,437 @@
...
+number-crunching.  While the jump from single-CPU to single-GPU is large, the
+jump from single-CPU to many-CPU or single-GPU to many-GPU is not as large as
+we would have liked.  For GPUs for example we got around a 2x speedup when we
+added 8x as many GPUs.
+
+At first one might think that this is because we're saturating disk read speeds.
+However two pieces of evidence go against that guess:
+
+-  NVIDIA folks familiar with my current hardware inform me that they're able to get
+   much more I/O when they're careful (TODO verify)
```
Comment (mrocklin) [inline _posts/2019-01-13-dask-cudf-first-steps.md]: cc @mt-jones

File: _posts/2019-01-13-dask-cudf-first-steps.md [hunk 3]
```diff
@@ -0,0 +1,437 @@
...
+
+Building Blocks: cuDF and Dask
+------------------------------
+
+Building a distributed GPU-backed dataframe is a large endeavor.
+Fortunately we're starting on a good foundation and
+can assemble much of this system from existing components:
+
+1.  The [cuDF](https://github.com/rapidsai/cudf) library aims to implement the
+    Pandas API on the GPU.  It gets good speedups (around 10x) on standard
```
Comment (harrism) [inline _posts/2019-01-13-dask-cudf-first-steps.md]: What is your comparison for 10x? One GPU vs. one CPU core? One GPU vs. multi-core CPU? DGX-2 vs. some number of Spark nodes on ETL benchmark?

File: _posts/2019-01-13-dask-cudf-first-steps.md [hunk 4]
```diff
@@ -0,0 +1,437 @@
...
+
+    ```python
+    import cudf  # looks and feels like Pandas, but runs on the GPU
+
+    df = cudf.read_csv('myfile.csv')
+    df = df[df.name == 'Alice']
+    df.groupby('id').value.mean()
+    ```
+
+    It is part of the growing [RAPIDS initiative](https://rapids.ai).
```
Comment (harrism) [inline _posts/2019-01-13-dask-cudf-first-steps.md]:
```suggestion
    cuDF is part of the growing [RAPIDS initiative](https://rapids.ai).
```

File: _posts/2019-01-13-dask-cudf-first-steps.md [hunk 5]
```diff
@@ -0,0 +1,437 @@
...
+    ```python
+    import dask.dataframe as dd  # looks and feels like Pandas, but runs in parallel
+
+    df = dd.read_csv('myfile.*.csv')
+    df = df[df.name == 'Alice']
+    df.groupby('id').value.mean().compute()
+    ```
+
+3.  The [Dask distributed task scheduler](https://distributed.dask.org)
+    provides general purpose parallel execution given complex task graphs.
```
Comment (harrism) [inline _posts/2019-01-13-dask-cudf-first-steps.md]:
```suggestion
    provides general-purpose parallel execution given complex task graphs.
```

File: _posts/2019-01-13-dask-cudf-first-steps.md [hunk 6]
```diff
@@ -0,0 +1,437 @@
...
+    It's good for adding multi-node computing into an existing codebase.
+
+Given these building blocks,
+our approach is to make cuDF look enough like Pandas that
```
Comment (harrism) [inline _posts/2019-01-13-dask-cudf-first-steps.md]:
```suggestion
our approach is to make the cuDF API close enough to Pandas that
```

File: _posts/2019-01-13-dask-cudf-first-steps.md [hunk 7]
```diff
@@ -0,0 +1,437 @@
...
+the Dask Dataframe algorithms can be reused.
```
Comment (harrism) [inline _posts/2019-01-13-dask-cudf-first-steps.md]:
```suggestion
we can reuse the Dask Dataframe algorithms.
```

File: _posts/2019-01-13-dask-cudf-first-steps.md [hunk 8]
```diff
@@ -0,0 +1,437 @@
...
+
+
+Benefits and Challenges to this approach
+----------------------------------------
+
+This approach has a few benefits:
+
+1.  We get to reuse the parallel algorithms found in Dask Dataframe originally designed for Pandas.
+
+    (these were hard enough to get right the first time)
```
Comment (harrism) [inline _posts/2019-01-13-dask-cudf-first-steps.md]:
```suggestion

```

Comment (harrism) [reply _posts/2019-01-13-dask-cudf-first-steps.md]: Seems unnecessary -- smart readers know that reusing good algorithms/implementations is important. If you keep it, rewrite it as a sentence; I don't think it needs to be parenthetical.

File: _posts/2019-01-13-dask-cudf-first-steps.md [hunk 9]
```diff
@@ -0,0 +1,437 @@
...
+
+2.  It consolidates the development effort behind a single codebase so that
```
Comment (harrism) [inline _posts/2019-01-13-dask-cudf-first-steps.md]:
```suggestion
2.  It consolidates the development effort within a single codebase so that
```

File: _posts/2019-01-13-dask-cudf-first-steps.md [hunk 10]
```diff
@@ -0,0 +1,437 @@
...
+    future effort spent on CPU Dataframes benefits GPU Dataframes and vice
+    versa.  Maintenance costs are shared.
+
+3.  By building code that works simultaneously with two Pandas implementations
```
Comment (harrism) [inline _posts/2019-01-13-dask-cudf-first-steps.md]:
```suggestion
3.  By building code that works equally with two DataFrame implementations
```

Comment (harrism) [reply _posts/2019-01-13-dask-cudf-first-steps.md]: Shouldn't call cuDF a Pandas implementation.

File: _posts/2019-01-13-dask-cudf-first-steps.md [hunk 11]
```diff
@@ -0,0 +1,437 @@
...
+    (CPU and GPU) we start to establish conventions and protocols that will
```
Comment (harrism) [inline _posts/2019-01-13-dask-cudf-first-steps.md]:
```suggestion
    (CPU and GPU) we establish conventions and protocols that will
```

File: _posts/2019-01-13-dask-cudf-first-steps.md [hunk 12]
```diff
@@ -0,0 +1,437 @@
...
+    make it easy for other projects to do the same, either with these two
```
Comment (harrism) [inline _posts/2019-01-13-dask-cudf-first-steps.md]:
```suggestion
    make it easier for other projects to do the same, either with these two
```

File: _posts/2019-01-13-dask-cudf-first-steps.md [hunk 13]
```diff
@@ -0,0 +1,437 @@
...
+    Pandas-like libraries, or with future Pandas-like libraries.
+
+    We start forcing the ecosystem to consider supporting Pandas-like
```
Comment (harrism) [inline _posts/2019-01-13-dask-cudf-first-steps.md]:
```suggestion
    This approach also aims to demonstrate that the ecosystem should support Pandas-like
```

Comment (harrism) [reply _posts/2019-01-13-dask-cudf-first-steps.md]: Forcing sounds... aggressive.

File: _posts/2019-01-13-dask-cudf-first-steps.md [hunk 14]
```diff
@@ -0,0 +1,437 @@
...
+    libraries, rather than just the single Pandas library.  For example if
```
Comment (harrism) [inline _posts/2019-01-13-dask-cudf-first-steps.md]:
```suggestion
    libraries, rather than just Pandas.  For example, if
```

File: _posts/2019-01-13-dask-cudf-first-steps.md [hunk 15]
```diff
@@ -0,0 +1,437 @@
...
+    (when?) the Arrow library develops a computational system then we'll be in
+    a better place to roll that in as well.
+
+4.  When doing any refactor we tend to clean up existing code.
+
+    For example to make dask dataframe ready for a new GPU Parquet reader
```
Comment (harrism) [inline _posts/2019-01-13-dask-cudf-first-steps.md]:
```suggestion
    For example, to make dask dataframe ready for a new GPU Parquet reader
```

File: _posts/2019-01-13-dask-cudf-first-steps.md [hunk 16]
```diff
@@ -0,0 +1,437 @@
...
+    we end up [refactoring and simplifying our Parquet I/O logic](https://github.com/dask/dask/pull/4336).
+
+But also some drawbacks:
```
Comment (harrism) [inline _posts/2019-01-13-dask-cudf-first-steps.md]:
```suggestion
The approach also has some drawbacks. Namely, it places API pressure on cuDF to match Pandas so:
```

File: _posts/2019-01-13-dask-cudf-first-steps.md [hunk 17]
```diff
@@ -0,0 +1,437 @@
...
+
+1.  This places API pressure on cuDF to match Pandas so...
```
Comment (harrism) [inline _posts/2019-01-13-dask-cudf-first-steps.md]:
```suggestion

```

File: _posts/2019-01-13-dask-cudf-first-steps.md [hunk 18]
```diff
@@ -0,0 +1,437 @@
...
+
+2.  Slight differences in API now cause larger problems
```
Comment (harrism) [inline _posts/2019-01-13-dask-cudf-first-steps.md]:
```suggestion
1. Slight differences in API now cause larger problems, such as these:
```

File: _posts/2019-01-13-dask-cudf-first-steps.md [hunk 19]
```diff
@@ -0,0 +1,437 @@
...
+
+    -  [Join column ordering differs rapidsai/cudf #251](https://github.com/rapidsai/cudf/issues/251)
+    -  [Groupby aggregation column ordering differs rapidsai/cudf #483](https://github.com/rapidsai/cudf/issues/483#issuecomment-453218151)
+
+3.  cuDF has some pressure on it to repeat what some believe to be mistakes in
```
Comment (harrism) [inline _posts/2019-01-13-dask-cudf-first-steps.md]:
```suggestion
2.  cuDF has some pressure on it to repeat what some believe to be mistakes in
```

File: _posts/2019-01-13-dask-cudf-first-steps.md [hunk 20]
```diff
@@ -0,0 +1,437 @@
...
+    the Pandas API.
+
+    For example, cuDF today supports missing values arguably more sensibly than
+    does Pandas.  Should cuDF have to revert to the old way of doing things
```
Comment (harrism) [inline _posts/2019-01-13-dask-cudf-first-steps.md]:
```suggestion
    Pandas.  Should cuDF have to revert to the old way of doing things
```

File: _posts/2019-01-13-dask-cudf-first-steps.md [hunk 21]
```diff
@@ -0,0 +1,437 @@
...
+    just to match Pandas semantics?  Dask Dataframe will probably have to need
```
Comment (harrism) [inline _posts/2019-01-13-dask-cudf-first-steps.md]:
```suggestion
    just to match Pandas semantics?  Dask Dataframe will probably need
```

File: _posts/2019-01-13-dask-cudf-first-steps.md [hunk 22]
```diff
@@ -0,0 +1,437 @@
...
+    to grow more flexible here to handle evolution and small differences
```
Comment (harrism) [inline _posts/2019-01-13-dask-cudf-first-steps.md]:
```suggestion
    to be more flexible in order to handle evolution and small differences
```

File: _posts/2019-01-13-dask-cudf-first-steps.md [hunk 23]
```diff
@@ -0,0 +1,437 @@
...
+    in semantics.
+
+
+Alternatives
+------------
+
+We could also write a new dask-dataframe-style project around cuDF that was
```
Comment (harrism) [inline _posts/2019-01-13-dask-cudf-first-steps.md]:
```suggestion
We could also write a new dask-dataframe-style project around cuDF that deviates
```

File: _posts/2019-01-13-dask-cudf-first-steps.md [hunk 24]
```diff
@@ -0,0 +1,437 @@
...
+more able to deviate from the Pandas/Dask Dataframe API.  Until recently this
```
Comment (harrism) [inline _posts/2019-01-13-dask-cudf-first-steps.md]:
```suggestion
from the Pandas/Dask Dataframe API.  Until recently this
```

File: _posts/2019-01-13-dask-cudf-first-steps.md [hunk 25]
```diff
@@ -0,0 +1,437 @@
...
+has actually been the approach, and the
+[dask-cudf](https://github.com/rapidsai/dask-cudf) project did exactly this.
+This was probably a good choice early on to get started and prototype things.
+The project was able to implement a wide range of functionality including
+groupby-aggregations, joins, and so on using
+[dask delayed](https://docs.dask.org/en/latest/delayed.html).
+
+We're redoing this now on top of dask dataframe though, which means that we're
+losing some functionality that dask-cudf had before, but hopefully the
```
Comment (harrism) [inline _posts/2019-01-13-dask-cudf-first-steps.md]:
```suggestion
losing some functionality that dask-cudf already had, but hopefully the
```

File: _posts/2019-01-13-dask-cudf-first-steps.md [hunk 26]
```diff
@@ -0,0 +1,437 @@
...
+          alt="Dask + CUDA + cuDF JupyterLab environment"
+          width="70%">
+   </a>
+
+-  Our `df` object is actually just a normal Dask Dataframe.  We didn't have to
+   write new `__repr__`, `__add__`, or `.sum()` implementations, and probably
+   many functions we didn't think about work well today (though also many
+   don't).
+
+-  We're tightly integrated and more connected to other systems. For example if
```
Comment (harrism) [inline _posts/2019-01-13-dask-cudf-first-steps.md]:
```suggestion
-  We're tightly integrated and more connected to other systems. For example, if
```

File: _posts/2019-01-13-dask-cudf-first-steps.md [hunk 27]
```diff
@@ -0,0 +1,437 @@
...
+   we wanted to convert our dask-cudf-dataframe to a dask-pandas-dataframe then
+   we would use the standard cuDF function for this:
```
Comment (harrism) [inline _posts/2019-01-13-dask-cudf-first-steps.md]:
```suggestion
   we would just use the cuDF `to_pandas` function:
```

File: _posts/2019-01-13-dask-cudf-first-steps.md [hunk 28]
```diff
@@ -0,0 +1,437 @@
...
+
+   ```python
+   df = df.map_partitions(cudf.DataFrame.to_pandas)
+   ```
+
+   We don't have to write anything special like a separate `.to_dask_dataframe`
+   method nor handle other special cases.
```
Comment (harrism) [inline _posts/2019-01-13-dask-cudf-first-steps.md]:
```suggestion
   method or handle other special cases.
```

File: _posts/2019-01-13-dask-cudf-first-steps.md [hunk 29]
```diff
@@ -0,0 +1,437 @@
...
+
+   The parallelism of Dask should be orthogonal to the choice of CPU/GPU.
```
Comment (harrism) [inline _posts/2019-01-13-dask-cudf-first-steps.md]:
```suggestion
   Dask parallelism is orthogonal to the choice of CPU or GPU.
```

File: _posts/2019-01-13-dask-cudf-first-steps.md [hunk 30]
```diff
@@ -0,0 +1,437 @@
...
+
+-  It's easy to switch hardware.  By avoiding separate `dask-cudf` code paths
+   it's easier to add cuDF to an existing Dask+Pandas codebase if we get a GPU,
```
Comment (harrism) [inline _posts/2019-01-13-dask-cudf-first-steps.md]:
```suggestion
   it's easier to add cuDF to an existing Dask+Pandas codebase to run on GPUs,
```

Comment (harrism) [reply _posts/2019-01-13-dask-cudf-first-steps.md]: "if we get a GPU" sounds weird. Who is we, and why do we have to share one GPU? :)

File: _posts/2019-01-13-dask-cudf-first-steps.md [hunk 31]
```diff
@@ -0,0 +1,437 @@
...
+   or remove cuDF and use Pandas if we want our code to be runnable by people
```
Comment (harrism) [inline _posts/2019-01-13-dask-cudf-first-steps.md]:
```suggestion
   or to remove cuDF and use Pandas if we want our code to be runnable without GPUs.
```

File: _posts/2019-01-13-dask-cudf-first-steps.md [hunk 32]
```diff
@@ -0,0 +1,437 @@
...
+   who don't have GPUs.
```
Comment (harrism) [inline _posts/2019-01-13-dask-cudf-first-steps.md]:
```suggestion

```

File: _posts/2019-01-13-dask-cudf-first-steps.md [hunk 33]
```diff
@@ -0,0 +1,437 @@
...
+
+   We'll see more examples of this in the scaling section below.
```
Comment (harrism) [inline _posts/2019-01-13-dask-cudf-first-steps.md]:
```suggestion
   There are more examples of this in the scaling section below.
```

File: _posts/2019-01-13-dask-cudf-first-steps.md [hunk 34]
```diff
@@ -0,0 +1,437 @@
...
+    (df = dd.read_csv('few-large/*.csv')
+            .repartition(npartitions=100)
+            .to_csv('many-small/*.csv', index=False))
+    ```
+
+    (See [rapidsai/cudf #568](https://github.com/rapidsai/cudf/issues/568))
+
+2.  Many operations that used to work in dask-cudf like groupby-aggregations
+    and joins no longer work.  We're going to need to slightly modify many cuDF
+    APIs over the next couple months to more closely match their Pandas
```
Comment (harrism) [inline _posts/2019-01-13-dask-cudf-first-steps.md]:
```suggestion
    APIs over the next couple of months to more closely match their Pandas
```

File: _posts/2019-01-13-dask-cudf-first-steps.md [hunk 35]
```diff
@@ -0,0 +1,437 @@
...
+    equivalents.
+
+3.  I ran the timing cell twice because it currently takes a few seconds to
+    `import cudf` today.
+    [rapidsai/cudf #627](https://github.com/rapidsai/cudf/issues/627)
```
Comment (harrism) [inline _posts/2019-01-13-dask-cudf-first-steps.md:266]: Anything that uses Numba JIT may also have JIT overhead. And there's also always overhead the first time you run a CUDA kernel, so some things will always be faster the second time you run them.

File: _posts/2019-01-13-dask-cudf-first-steps.md [hunk 36]
```diff
@@ -0,0 +1,437 @@
...
+
+4.  We had to make Dask Dataframe a bit more flexible and assume less about its
```
Comment (harrism) [inline _posts/2019-01-13-dask-cudf-first-steps.md:268]: You say this is a bad thing, but it's exactly what you were happy about "forcing the ecosystem to do" above... Maybe rephrase to be more consistent in your message?

File: _posts/2019-01-13-dask-cudf-first-steps.md [hunk 37]
```diff
@@ -0,0 +1,437 @@
...
+    constituent dataframes being exactly Pandas dataframes.  (see
+    [dask/dask #4359](https://github.com/dask/dask/pull/4359) and
+    [dask/dask #4375](https://github.com/dask/dask/pull/4375) for examples).
+    I suspect that there will by many more small changes like
+    these necessary in the future.
+
+These problems are representative of dozens more similar issues.  They are are
```
Comment (harrism) [inline _posts/2019-01-13-dask-cudf-first-steps.md]:
```suggestion
These problems are representative of dozens more similar issues.  They are
```

File: _posts/2019-01-13-dask-cudf-first-steps.md [hunk 38]
```diff
@@ -0,0 +1,437 @@
...
+Near Term Schedule
+------------------
+
+The RAPIDS group is currently busy working to release 0.5, which includes some
+of the fixes necessary to run the example above, and also many unrelated
+stability improvements.  This will probably keep them busy for a week or two
+during which I don't expect to see much Dask + cuDF work going on other than
+planning.
+
+After that, and for the next couple months this seems to be their priority, so
```
Comment (harrism) [inline _posts/2019-01-13-dask-cudf-first-steps.md]:
```suggestion
After that, Dask parallelism support will be a top priority, so
```

File: _posts/2019-01-13-dask-cudf-first-steps.md [hunk 39]
```diff
@@ -0,0 +1,437 @@
...
+
+Analysis
+--------
+
+First, lets appreciate again that it's easy to test a wide variety of
+architectures using this setup because of the Pandas API compatibility between
+all of the different projects.  We're seeing a wide range of performance (40x
+span) across a variety of different hardware with a wide range of cost points.
+
+Second lets note that this problem scales less well than our
```
Comment (harrism) [inline _posts/2019-01-13-dask-cudf-first-steps.md]:
```suggestion
Second, note that this problem scales less well than our
```

File: _posts/2019-01-13-dask-cudf-first-steps.md [hunk 40]
```diff
@@ -0,0 +1,437 @@
...
+  </tbody>
+</table>
+
+*The bandwidth numbers were computed by noting that the data was around 10 GB on disk*
+
+
+Analysis
+--------
+
+First, lets appreciate again that it's easy to test a wide variety of
```
Comment (harrism) [inline _posts/2019-01-13-dask-cudf-first-steps.md]:
```suggestion
First, I want to emphasize again that it's easy to test a wide variety of
```

File: _posts/2019-01-13-dask-cudf-first-steps.md (hunk 2 above)
Comment (harrism) [inline _posts/2019-01-13-dask-cudf-first-steps.md]: rephrase "more I/O" to be clearer. Do you mean "higher I/O throughput"?

File: _posts/2019-01-13-dask-cudf-first-steps.md [hunk 41]
```diff
@@ -0,0 +1,437 @@
...
+(without us doing more work that is).
+
+
+### Come help!
+
+If the work above sounds interesting to you then come help!
+There is a lot of low-hanging and high impact work to do.
+
+If you're interested in being paid to focus more on these topics, then consider
+applying for a job.  The NVIDIA corporation is hiring around the use of Dask
```
Comment (harrism) [inline _posts/2019-01-13-dask-cudf-first-steps.md]:
```suggestion
applying for a job.  NVIDIA's RAPIDS team is looking to hire engineers for Dask development with GPUs.
```

File: _posts/2019-01-13-dask-cudf-first-steps.md [hunk 42]
```diff
@@ -0,0 +1,437 @@
...
+with GPUs.
```
Comment (harrism) [inline _posts/2019-01-13-dask-cudf-first-steps.md]:
```suggestion

```

File: _posts/2019-01-13-dask-cudf-first-steps.md [hunk 43]
```diff
@@ -0,0 +1,437 @@
...
+
+-  [Senior Library Software Engineer - RAPIDS](https://nvidia.wd5.myworkdayjobs.com/en-US/NVIDIAExternalCareerSite/job/US-TX-Austin/Senior-Library-Software-Engineer---RAPIDS_JR1919608-1)
+
+That's a fairly generic posting.  If you're interested the posting doesn't seem
```
Comment (harrism) [inline _posts/2019-01-13-dask-cudf-first-steps.md]: Suggest dropping this ending and just say "and other data analytics library development projects" or something similar in the previous sentence.

**Review by harrism — APPROVED** [review]
Great post. My comments are mostly just minor fix-ups for style and flow. 

My only complaint is that you rely on "we" a lot, and it starts to sound unnatural after a while. My opinion is that It's you writing the post, use "I" where appropriate, and "we" when talking about team efforts. I find this sounds more personal and natural and avoids the barrier that "royal we" sometimes sets up.

File: _posts/2019-01-13-dask-cudf-first-steps.md (hunk 3 above)
Comment (mrocklin) [reply _posts/2019-01-13-dask-cudf-first-steps.md]: Nothing serious, just a few simple spot checks.  I'll drop the claim.

File: _posts/2019-01-13-dask-cudf-first-steps.md (hunk 43 above)
Comment (mrocklin) [reply _posts/2019-01-13-dask-cudf-first-steps.md]: OK, will do.

**Review by mrocklin — COMMENTED** [review]
Thanks for the suggestions @harrism .  Those were great.  I've gone ahead and committed them.

File: _posts/2019-01-13-dask-cudf-first-steps.md (hunk 1 above)
Comment (mrocklin) [reply _posts/2019-01-13-dask-cudf-first-steps.md:277]: It seems to still be populated

File: _posts/2019-01-13-dask-cudf-first-steps.md (hunk 3 above)
Comment (harrism) [reply _posts/2019-01-13-dask-cudf-first-steps.md]: I didn't want you to drop it! On the contrary, I just want you to make the claim stronger. We demonstrated (conservatively) more than 50x on ETL (Vs. Spark) at GTCE.

Comment (mrocklin) [reply _posts/2019-01-13-dask-cudf-first-steps.md]: Yeah, it's a complex topic though.  Lets address it with specifics another time.

**Comment by mrocklin** [comment]
OK, I plan to release this tomorrow morning if there are no objections
